File: tools/town_generator/src/wards/Market.cpp
#include "town_generator/wards/Market.h"
#include "town_generator/building/City.h"
#include "town_generator/utils/Random.h"
#include <cmath>

namespace town_generator {
namespace wards {

geom::Polygon Market::getAvailable() {
    // Faithful to mfcg.js Market.getAvailable (lines 652-677)
    // Market fills the entire patch, only insets for canal edges
    if (!patch || !model) return patch ? patch->shape : geom::Polygon();

    size_t len = patch->shape.length();

    // Start with zero insets for all edges (market fills entire space)
    std::vector<double> edgeInsets(len, 0.0);

    // Per-vertex canal exclusions (for corner handling)
    std::vector<double> vertexExclusions(len, 0.0);

    // Check each edge for canal type
    for (size_t i = 0; i < len; ++i) {
        const geom::Point& v0 = patch->shape[i];
        const geom::Point& v1 = patch->shape[(i + 1) % len];

        // Check if edge is on a canal - use canal width / 2
        for (const auto& canal : model->canals) {
            if (canal->containsEdge(v0, v1)) {
                edgeInsets[i] = canal->width / 2.0;
                break;
            }

            // Check canal vertex exclusions
            double canalWidth = canal->getWidthAtVertex(v0);
            if (canalWidth > 0) {
                vertexExclusions[i] = std::max(vertexExclusions[i], canalWidth / 2.0);
                // If this is the start of the canal, add extra offset
                if (!canal->course.empty() && canal->course[0] == v0) {
                    vertexExclusions[i] += ALLEY;
                }
            }
        }
    }

    // Apply vertex exclusions to adjacent edges
    for (size_t i = 0; i < len; ++i) {
        if (vertexExclusions[i] > 0) {
            edgeInsets[i] = std::max(edgeInsets[i], vertexExclusions[i]);
            size_t prevIdx = (i + len - 1) % len;
            edgeInsets[prevIdx] = std::max(edgeInsets[prevIdx], vertexExclusions[i]);
        }
    }

    return patch->shape.shrink(edgeInsets);
}

void Market::createGeometry() {
    // Faithful to mfcg.js Market.createGeometry (lines 632-651)
    if (!patch) return;

    utils::Random::reset(patch->seed);

    geometry.clear();

    // Get available space (uses zero insets, fills whole patch)
    space = getAvailable();
    if (space.length() < 3) return;

    // 60% chance for rectangular statue/monument, else octagonal fountain
    bool statue = utils::Random::boolVal(0.6);
    // If statue, always offset; otherwise 30% chance for offset
    bool offset = statue || utils::Random::boolVal(0.3);

    geom::Point c, d;  // Longest edge endpoints
    size_t longestIdx = 0;

    if (statue || offset) {
        // Find longest edge of the space (not patch->shape)
        double maxLen = -1.0;
        for (size_t i = 0; i < space.length(); ++i) {
            const geom::Point& p0 = space[i];
            const geom::Point& p1 = space[(i + 1) % space.length()];
            double len = geom::Point::distance(p0, p1);
            if (len > maxLen) {
                maxLen = len;
                longestIdx = i;
            }
        }
        c = space[longestIdx];
        d = space[(longestIdx + 1) % space.length()];
    }

    geom::Polygon monument;
    if (statue) {
        // Rectangular statue/monument aligned to longest edge
        monument = geom::Polygon::rect(
            1.0 + utils::Random::floatVal(),
            1.0 + utils::Random::floatVal()
        );
        // Rotate to align with longest edge
        geom::Point edgeDir = d.subtract(c);
        double angle = std::atan2(edgeDir.y, edgeDir.x);
        monument.rotate(angle);
    } else {
        // Octagonal fountain (regular(8, radius))
        monument = geom::Polygon::regular(8, 1.0 + utils::Random::floatVal());
    }

    // Position: at centroid of space, or offset toward longest edge
    geom::Point centroid = space.centroid();
    if (offset) {
        // Offset toward midpoint of longest edge
        geom::Point edgeMid((c.x + d.x) / 2.0, (c.y + d.y) / 2.0);
        double t = 0.2 + utils::Random::floatVal() * 0.4;
        geom::Point pos(
            centroid.x + (edgeMid.x - centroid.x) * t,
            centroid.y + (edgeMid.y - centroid.y) * t
        );
        monument.offset(pos);
    } else {
        monument.offset(centroid);
    }

    geometry.push_back(monument);
}

} // namespace wards
} // namespace town_generator

File: tools/town_generator/reference/mfcg-clean/wards/Market.js
/**
 * Market.js - Market square ward
 *
 * Represents the central market plaza of the city.
 */

import { Point } from '../core/Point.js';
import { Random } from '../core/Random.js';
import { Ward } from './Ward.js';
import { PolygonUtils } from '../geometry/PolygonUtils.js';
import { GeomUtils } from '../geometry/GeomUtils.js';

/**
 * Market square ward
 */
export class Market extends Ward {
    /**
     * Create a market ward
     * @param {import('../model/City.js').City} model
     * @param {import('../model/Cell.js').Cell} patch
     */
    constructor(model, patch) {
        super(model, patch);

        /** @type {Point[]} */
        this.space = [];

        /** @type {Point[]} */
        this.monument = [];
    }

    /**
     * Get available area (market-specific)
     * @returns {Point[]|null}
     */
    getAvailable() {
        const shape = this.patch.shape;
        const n = shape.length;

        const cornerRadii = new Array(n).fill(0);
        const edgeInsets = [];

        let i = 0;
        for (const edge of this.patch.face.edges()) {
            // Check for canal at vertex
            for (const canal of this.model.canals) {
                const courseEdge = canal.course.find(e => e.origin === edge.origin);
                if (courseEdge) {
                    cornerRadii[i] = canal.width / 2;
                    if (edge.origin === canal.course[0].origin) {
                        cornerRadii[i] += 1.2;
                    }
                }
            }

            // Edge inset based on type
            if (edge.data === 4) { // CANAL
                edgeInsets.push(this.model.canals[0]?.width / 2 || 2);
            } else {
                edgeInsets.push(0);
            }
            i++;
        }

        return Ward.inset(shape, edgeInsets, cornerRadii);
    }

    /**
     * Create market geometry
     */
    createGeometry() {
        Random.restore(this.patch.seed);

        this.space = this.getAvailable();
        if (!this.space) {
            this.space = this.patch.shape.slice();
        }

        // Decide monument type
        const hasRect = Random.bool(0.6);
        const hasMonument = hasRect || Random.bool(0.3);

        let monumentPos = null;
        let monumentDir = null;

        if (hasMonument) {
            // Find longest edge for orientation
            const longestIdx = PolygonUtils.longestEdge(this.space);
            const p1 = this.space[longestIdx];
            const p2 = this.space[(longestIdx + 1) % this.space.length];

            monumentPos = GeomUtils.lerp(p1, p2);
            monumentDir = p2.subtract(p1);
        }

        if (hasRect) {
            // Rectangular monument (fountain, statue base)
            const width = 1 + Random.next();
            const height = 1 + Random.next();
            this.monument = PolygonUtils.rectangle(width, height);

            const angle = Math.atan2(monumentDir.y, monumentDir.x);
            PolygonUtils.rotate(this.monument, angle);
        } else {
            // Circular monument (well, column)
            const radius = 1 + Random.next();
            this.monument = PolygonUtils.regular(8, radius);
        }

        // Position monument
        const center = PolygonUtils.centroid(this.space);
        if (monumentPos) {
            // Between center and edge
            const pos = GeomUtils.lerp(center, monumentPos, 0.2 + Random.next() * 0.4);
            PolygonUtils.translateBy(this.monument, pos);
        } else {
            PolygonUtils.translateBy(this.monument, center);
        }
    }
}
